0000-ilp-over-http.md

```diff
@@ -0,0 +1,59 @@
+---
+title: ILP Over HTTP
+draft: 1
+---
+
+# ILP Over HTTP
+> A bilateral communication protocol for server-to-server connections
+
+## Motivation
+
+Scaling Interledger infrastructure to handle large volumes of ILP packets requires horizontally scaling connectors. Using HTTP for bilateral communication enables service providers to leverage standard tools and services for hosting, load balancing, Distributed Denial of Service (DDoS) protection, and monitoring.
+
+## Overview
+
+In an ILP Over HTTP connection, both peers run HTTP servers with accessible HTTPS endpoints. When peering, the peers exchange their respective URLs, authentication tokens or TLS certificates, ILP addresses, and settlement-related details.
+
+Each ILP Prepare packet is sent as the body of an HTTP request to the peer's server endpoint. ILP Fulfill or Reject packets are returned as the body of the HTTP response.
+
+## Specification
+
+This is a minimal protocol built on HTTP. HTTP/2 is HIGHLY RECOMMENDED for performance reasons, although HTTP/1.1 MAY also be used. Implementations SHOULD support HTTP version negotiation via Application Protocol Negotiation (ALPN).
+
+### Authentication
+
+Peers MAY use any standard HTTP authentication mechanism to authenticate incoming requests. Bearer Tokens or TLS Client Certificates are RECOMMENDED.
+
+### Request
+
+```http
+POST /ilp HTTP/x.x
+Accept: application/ilp+octet-stream
+Content-Type: application/ilp+octet-stream
+Authorization: Bearer zxcljvoizuu09wqqpowipoalksdflksjdgxclvkjl0s909asdf
+Host: connector.example
+ILP-Destination: example.usd.connector.account
+
+< Body: Binary OER-Encoded ILP Prepare Packet >
+```
+
+**Field Details:**
+
+- **Path** - A connector MAY specify any HTTP path for their peer to send ILP packets to.
+- **ILP-Destination Header** - (Optional) It is RECOMMENDED to set this header to the destination ILP address from the ILP packet. This enables standard HTTP load balancers to route packets to particular connector instances without parsing the ILP packet.
+- **Body** - ILP Packet encoded using OER, as specified in [RFC 27: Interledger Protocol V4](./0027-interledger-protocol-4/0027-interledger-protocol-4.md).
+
+### Response
+
+```http
+HTTP/x.x 200 OK
+Content-Type: application/ilp+octet-stream
+
+< Body: Binary OER-Encoded ILP Fulfill or Reject Packet >
+```
+
+All ILP Packets MUST be returned with the HTTP status code `200: OK`. An endpoint MAY return standard HTTP errors, including but not limited to: a malformed or unauthenticated request, rate limiting, or an unresponsive upstream service. Connectors SHOULD either retry the request, if applicable, or relay an ILP Reject packet back to the original sender with an appropriate [Final or Temporary error code](./0027-interledger-protocol-4/0027-interledger-protocol-4#error-codes).
```

Comment:
> Connectors SHOULD either retry the request, if applicable, or relay an ILP Reject packet back to the original sender

I think it would be worthwhile to clarify (non-normatively) what some of these applicable cases _might_ be. Probably this means just suggesting how to handle ambiguous HTTP error-code responses, which I think limited to the `5xx` family. 

I.e., cases where it's ambiguous what the receiving peer did with the ILP-prepare packet. For example, an HTTP 500 doesn't tell me if my peer meant to respond with a `FULFILL` but blew-up before being able to do so; or...if it instead meant to `REJECT`, but fell over before it could say that.

Reply:
It might actually be OK to say that a `5xx` error code should trigger a `T00` error, and retries should never be done at this layer. STREAM already handles retrying so we shouldn't put unnecessary work onto the connector.

Reply:
Clarification: are HTTP error codes ever sent *in lieu of* an ILP reject (e.g. in the case the request was authenticated, got through the load balancer, and the PREPARE was correctly formed)?

Reply:
Possible case may be that the HTTP module on the connector is overloaded and is simply responding to all requests with a 503. I agree with Ben that the safest thing to do is always handle a 5XX response by triggering a T00 error (or simply letting the request time out?).

Reply:
What about T01: Peer Unreachable, T02: Peer Busy, or T05: Rate Limited? Those sound like they _could_ be translations of HTTP 5xx errors.

That said, it would be easiest to implement if you just translate all 5xx errors into T00 ILP errors.